### PR TITLE
flow/operator: Allow prometheus.operator.*monitor components to reference cluster secrets

### LINKED ISSUE #3749: Flow: Support Secrets referenced by Monitoring CRDs.

### LINKED ISSUE #4067: x509 Unknown Authority when scraping ServiceMonitor objects
### What's wrong?

I'm trying to setup grafana agent to discover servicemonitor objects to retrieve metrics from the k8s apiserver.

I'm encountering a `failed to verify certificate: x509: certificate signed by unknown authority ` in the agent logs.

As it is said in the documentation, the agent should take the default kubernetes configuration to talk to the apiserver.
I tried to specify the configuration explicitly in both the agent config and the servicemonitor, set the insecure_skip_verify flag to true but the result is the same each time. It seems like it is not taken into account (even though it is shown in the UI)

Note that if I try to do a manual curl using the same certificate, it works which makes me think it comes from the agent itself.

Please see https://grafana.slack.com/archives/C01050C3D8F/p1686060128742699 for the initial discussion.

### Steps to reproduce

- Deploy the Grafana Agent on a K8s cluster
- Deploy the apiserver ServiceMonitor
- Go to the Grafana Agent UI and check the status of the apiserver service monitor

### System information

Kubernetes v1.25.9

### Software version

Grafana Agent v0.34.0-rc.1

### Configuration

```text
# config.river
prometheus.operator.servicemonitors "servicemonitors" {
  forward_to = [otelcol.receiver.prometheus.metrics.receiver]

  client {
    api_server = "https://kubernetes.default.svc"
    bearer_token_file = "/var/run/secrets/kubernetes.io/serviceaccount/token"
    tls_config {
      ca_file = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
      insecure_skip_verify = true
    }
  }
}

otelcol.receiver.prometheus "metrics" {
  output {
    metrics = [otelcol.exporter.prometheus.mimir.input]
  }
}

otelcol.exporter.prometheus "mimir" {
  forward_to = [prometheus.remote_write.mimir.receiver]
}

prometheus.remote_write "mimir" {
  external_labels = {
    cluster = "dev-core",
  }
  endpoint {
    url = "http://mimir-nginx/api/v1/push"
  }
}

# apiserver-servicemonitor.yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: apiserver
  namespace: monitoring
spec:
  endpoints:
  - metricRelabelings:
    - action: drop
      regex: apiserver_request_duration_seconds_bucket;(0.15|0.2|0.3|0.35|0.4|0.45|0.6|0.7|0.8|0.9|1.25|1.5|1.75|2|3|3.5|4|4.5|6|7|8|9|15|25|40|50)
      sourceLabels:
      - __name__
      - le
    tlsConfig:
      caFile: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    port: https
    scheme: https
    interval: 30s
    bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
  jobLabel: component
  namespaceSelector:
    matchNames:
    - default
  selector:
    matchLabels:
      component: apiserver
      provider: kubernetes
```

### Logs

```
ts=2023-06-06T14:52:00.304859547Z component=otelcol.receiver.prometheus.metrics level=warn msg="Failed to scrape Prometheus endpoint" scrape_timestamp=1686063120303 target_labels="{__name__=\"up\", endpoint=\"https\", instance=\"10.69.94.123:6443\", job=\"apiserver\", namespace=\"default\", service=\"kubernetes\"}"
ts=2023-06-06T14:52:00.991405398Z component=prometheus.operator.servicemonitors.servicemonitors level=debug scrape_pool=serviceMonitor/monitoring/apiserver/0 target=https://10.69.95.160:6443/metrics msg="Scrape failed" err="Get \"https://10.69.95.160:6443/metrics\": tls: failed to verify certificate: x509: certificate signed by unknown authority"
```

### PR DESCRIPTION
**Pull request by captncraig**
Fixes #3749 
Fixes #4067

Uses implementation described there. Namely we are fetching secrets on-demand when we reconcile the podmonitors. It will only react to secret changes when the resync triggers every 5 minutes, or a podmonitor changes.

- [x] Changelog
- [ ] Should resync time be configurable? 5 minutes is hardcoded now.
- [x] Update docs to reflect behavior and resync time
- [ ] Tests for cases where we fail to get secret.

### CODE REVIEW

File: component/prometheus/operator/configgen/config_gen_test.go [hunk 1]
```diff
@@ -153,54 +191,93 @@ func TestGenerateSafeTLSConfig(t *testing.T) {
...
 			assert.True(t, got.InsecureSkipVerify)
 			assert.Equal(t, tt.serverName, got.ServerName)
+			assert.Equal(t, tt.ca, got.CA)
+			assert.Equal(t, tt.cert, got.Cert)
+			assert.Equal(t, tt.key, got.Key)
+		})
+	}
+}
+
+func TestGenerateBasicAuth(t *testing.T) {
```
Comment (jcreixell) [inline component/prometheus/operator/configgen/config_gen_test.go:254]: let's add tests also for the other missing functions?

**Comment by jcreixell** [comment]
looks like it is going in the right direction, clean and simple

File: docs/sources/flow/reference/components/prometheus.operator.podmonitors.md [hunk 2]
```diff
@@ -16,6 +16,8 @@ labels:
 
 The default configuration assumes the agent is running inside a Kubernetes cluster, and uses the in-cluster config to access the Kubernetes API. It can be run from outside the cluster by supplying connection info in the `client` block, but network level access to pods is required to scrape metrics from them.
 
+PodMonitors may reference secrets for authenticating to targets to scrape them. In these cases, the secrets will be loaded and refreshed only when the PodMonitor itself is updated, or when this component refreshes its' internal state, which happens on a 5 minute refresh cycle.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/prometheus.operator.podmonitors.md:19]:
```suggestion
PodMonitors may reference secrets for authenticating to targets to scrape them. In these cases, the secrets are loaded and refreshed only when the PodMonitor itself is updated or when this component refreshes its' internal state, which happens on a 5-minute refresh cycle.
```

File: docs/sources/flow/reference/components/prometheus.operator.servicemonitors.md [hunk 3]
```diff
@@ -16,6 +16,8 @@ labels:
 
 The default configuration assumes the agent is running inside a Kubernetes cluster, and uses the in-cluster config to access the Kubernetes API. It can be run from outside the cluster by supplying connection info in the `client` block, but network level access to discovered endpoints is required to scrape metrics from them.
 
+ServiceMonitors may reference secrets for authenticating to targets to scrape them. In these cases, the secrets will be loaded and refreshed only when the ServiceMonitor itself is updated, or when this component refreshes its' internal state, which happens on a 5 minute refresh cycle.
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/prometheus.operator.servicemonitors.md:19]:
```suggestion
ServiceMonitors may reference secrets for authenticating to targets to scrape them. In these cases, the secrets are loaded and refreshed only when the ServiceMonitor is updated or when this component refreshes its' internal state, which happens on a 5-minute refresh cycle.
```

**Review by clayton-cornell — COMMENTED** [review]
Minor changes  to make the text more active.

**Review by clayton-cornell — COMMENTED** [review]
All good from a docs perspective.

**Comment by jcreixell** [comment]
@captncraig let me know when this is ready for a second review pass (I see some pending TODOs and unaddressed comments so haven't looked at it again yet)

**Review by jcreixell — APPROVED** [review]

**Pull request merged**